### DISCUSSION TITLE
Migration path from play 2.2.3 to latest

### DISCUSSION THREAD

**Discussion by javadogma** [playframework/playframework#12790] [General]
Hello,

I inherited a play framework project a few years ago that uses play 2.2.3 built with Scala 2.10.3 (running Java 1.8.0_112) and sbt 0.13.  From a project perspective, the IDE is MyEclipse and it somehow knows how to build the projects so I think sbt is somehow configured behind the scenes.

I'm looking for an outline, playbook, grimoire, of what the migration path looks like to get this project on the latest release of play, Scala and sbt. My hope is that it's not 2.2.3 -> 2.2.4 -> 2.3.1 -> ... -> latest and that MyEclipse doesn't bark afterward.

Thanks,
Nick.

**Comment by tsuyoshizawa** [comment]
Migration instructions are available on the official website.
Hardship is a problem that occurs in all frameworks.
Try to take it one step at a time.

- https://www.playframework.com/documentation/3.0.x/Migration23
- https://www.playframework.com/documentation/3.0.x/Migration24
- https://www.playframework.com/documentation/3.0.x/Migration25
- https://www.playframework.com/documentation/3.0.x/Migration26
- https://www.playframework.com/documentation/3.0.x/Migration27
- https://www.playframework.com/documentation/3.0.x/Migration28
- https://www.playframework.com/documentation/3.0.x/Migration29
- https://www.playframework.com/documentation/3.0.x/Migration30

**Reply by jducoeur** [reply]
Yeah, agreed.  You probably don't need to go through every point release, but you *should* migrate one major release at a time, and it extremely important to deal with the deprecation warnings along the way -- there were some significant architectural shifts in the 2.4 - 2.6 range that you probably can't skip over.  (Particularly having to do with how dependency injection works.)

I recommend interleaving the Scala evolutions along the way: while it's not quite a precise "this version of Play requires this version of Scala", there tend to be assumptions of which versions were normal at any given time.  (And 2.10 is long since dead.)  Note that the jump from Scala 2.13 -> 3.0 is much larger, and is optional -- lots of projects are still on 2.13, and it's expected to be supported indefinitely, so for a legacy project like this you may want to stop there, or at least put off the jump to Scala 3 until everything else is done.  Aside from that, the language evolution will probably be the easy part: the language has mostly stayed back-compatible, so it's likely that not much will need to change in the code.

sbt is the one that's likeliest to cause you trouble with your IDE.  I'll be honest: Eclipse is pretty rare in the Scala community these days, and I don't know of any well-supported plugins currently.  Most folks are using IntelliJ IDEA or Visual Studio Code with Metals instead.  (I was a big Eclipse devotee in the earlier days of Play, but it just slowly decayed to the point of being barely usable.)

It shouldn't be an eternal project, but expect it to take a while -- you're speed-running something like 15 years of platform evolution at all levels, so there have been a fair number of changes along the way.

**Reply by javadogma** [reply]
Thank you for the insight and direction. That was very helpful.
